Sources/HttpPipeline/NIO.swift

```diff
@@ -0,0 +1,143 @@
+    let reqPart = self.unwrapInboundIn(data)
+
+    switch reqPart {
+    case let .head(header):
+      self.request = URL(string: header.uri).map {
+        var req = URLRequest(url: $0)
+          |> \.httpMethod .~ method(from: header.method)
+          |> \.allHTTPHeaderFields .~ header.headers.reduce(into: [:]) { $0[$1.name] = $1.value }
+        let proto = req.value(forHTTPHeaderField: "X-Forwarded-Proto") ?? "http"
+        req.url = req.value(forHTTPHeaderField: "Host").flatMap {
+          URL(string: proto + "://" + $0 + header.uri)
+        }
+        return req
+      }
+    case let .body(bodyPart):
+      self.request = self.request.flatMap { req -> URLRequest? in
+        req.value(forHTTPHeaderField: "Content-Length")
+          .flatMap(Int.init)
+          .flatMap { bodyPart.getBytes(at: 0, length: $0) }
+          .map { req |> \.httpBody .~ Data($0) }
+      }
+    case .end:
+      guard let req = self.request else {
+        _ = ctx.channel.close()
+        return
+      }
+
+      let conn = self.middleware(connection(from: req)).perform()
+      let res = conn.response
+
+      let head = HTTPResponseHead(
+        version: .init(major: 1, minor: 1),
+        status: .init(statusCode: res.status.rawValue),
+        headers: .init(res.headers.map { ($0.name, $0.value) })
+      )
+      _ = ctx.channel.write(HTTPServerResponsePart.head(head))
+
+      var buffer = ctx.channel.allocator.buffer(capacity: res.body.count)
+      buffer.write(bytes: res.body)
+      _ = ctx.channel.write(HTTPServerResponsePart.body(.byteBuffer(buffer)))
+
+      _ = ctx.channel.writeAndFlush(HTTPServerResponsePart.end(nil)).then {
+        ctx.channel.close()
+      }
+    }
+  }
+}
+
+private func method(from method: HTTPMethod) -> String {
+  switch method {
+  case .GET: return "GET"
+  case .PUT: return "PUT"
+  case .ACL: return "ACL"
+  case .HEAD: return "HEAD"
+  case .POST: return "POST"
+  case .COPY: return "COPY"
+  case .LOCK: return "LOCK"
+  case .MOVE: return "MOVE"
+  case .BIND: return "BIND"
+  case .LINK: return "LINK"
+  case .PATCH: return "PATCH"
+  case .TRACE: return "TRACE"
+  case .MKCOL: return "MKCOL"
+  case .MERGE: return "MERGE"
+  case .PURGE: return "PURGE"
+  case .NOTIFY: return "NOTIFY"
+  case .SEARCH: return "SEARCH"
+  case .UNLOCK: return "UNLOCK"
+  case .REBIND: return "REBIND"
+  case .UNBIND: return "UNBIND"
+  case .REPORT: return "REPORT"
+  case .DELETE: return "DELETE"
+  case .UNLINK: return "UNLINK"
+  case .CONNECT: return "CONNECT"
+  case .MSEARCH: return "MSEARCH"
+  case .OPTIONS: return "OPTIONS"
+  case .PROPFIND: return "PROPFIND"
+  case .CHECKOUT: return "CHECKOUT"
+  case .PROPPATCH: return "PROPPATCH"
+  case .SUBSCRIBE: return "SUBSCRIBE"
+  case .MKCALENDAR: return "MKCALENDAR"
+  case .MKACTIVITY: return "MKACTIVITY"
+  case .UNSUBSCRIBE: return "UNSUBSCRIBE"
+  case let .RAW(value): return value
+  }
+}
+
+private func headers(from headers: HTTPHeaders) -> [String: String] {
```

Comment:
this doesn't look used anymore

Reply:
Whoops, yeah. Deleted on the branch I had goin before 4.1 but forgot to update this one.